engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java

```diff
@@ -913,106 +899,28 @@ private void saveMetaInfo() throws IOException
-        operatorStatus.latencyMA.add(adjustedEndWindowEmitTimestamp - upstreamMaxEmitTimestamp);
-      } else {
-        operatorStatus.latencyMA.add(0);
-        if (lastLatencyWarningTime < System.currentTimeMillis() - LATENCY_WARNING_THRESHOLD_MILLIS) {
-          LOG.warn("Latency calculation for this operator may not be correct because upstream end window timestamp is greater than this operator's end window timestamp: {} ({}) > {} ({}). Please verify that the system clocks are in sync in your cluster. You can also try tweaking the RPC_LATENCY_COMPENSATION_SAMPLES application attribute (currently set to {}).",
-                  upstreamMaxEmitTimestamp, upstreamMaxEmitTimestampOperator, adjustedEndWindowEmitTimestamp, oper, this.vars.rpcLatencyCompensationSamples);
-          lastLatencyWarningTime = System.currentTimeMillis();
-        }
-      }
-    }
-
-    if (oper.getOutputs().isEmpty()) {
-      // it is a leaf operator
-      leafOperators.add(oper);
-    }
-    else {
-      for (PTOperator.PTOutput output : oper.getOutputs()) {
-        for (PTOperator.PTInput input : output.sinks) {
-          if (input.target != null) {
-            PTOperator downStreamOp = input.target;
-            if (!endWindowStatsVisited.contains(downStreamOp)) {
-              calculateLatency(downStreamOp, endWindowStatsMap, endWindowStatsVisited, leafOperators);
-            }
-          }
-        }
+    CriticalPathInfo result = null;
+    List<PTOperator> leafOperators = plan.getLeafOperators();
+    for (PTOperator leafOperator : leafOperators) {
+      CriticalPathInfo cpi = new CriticalPathInfo();
+      findCriticalPathHelper(leafOperator, cpi);
+      if (result == null || result.latency < cpi.latency) {
+        result = cpi;
       }
     }
+    return result;
   }
-  /*
-   * returns cumulative latency
-   */
-  private long findCriticalPath(Map<Integer, EndWindowStats> endWindowStatsMap, Set<PTOperator> operators, LinkedList<Integer> criticalPath)
-  {
-    long maxEndWindowTimestamp = 0;
-    PTOperator maxOperator = null;
-    for (PTOperator operator : operators) {
-      EndWindowStats endWindowStats = endWindowStatsMap.get(operator.getId());
-      if (maxEndWindowTimestamp < endWindowStats.emitTimestamp) {
-        maxEndWindowTimestamp = endWindowStats.emitTimestamp;
-        maxOperator = operator;
-      }
-    }
-    if (maxOperator == null) {
-      return 0;
-    }
-    criticalPath.addFirst(maxOperator.getId());
-    OperatorStatus operatorStatus = maxOperator.stats;
 
-    operators.clear();
-    if (maxOperator.getInputs() == null || maxOperator.getInputs().isEmpty()) {
-      return operatorStatus.latencyMA.getAvg();
-    }
-    for (PTOperator.PTInput input : maxOperator.getInputs()) {
-      if (null != input.source.source) {
-        operators.add(input.source.source);
-      }
+  private void findCriticalPathHelper(PTOperator operator, CriticalPathInfo cpi)
+  {
+    cpi.latency += operator.stats.getLatencyMA();
```

Comment:
@gauravgopi123 Yes, you are correct.  We actually need to change this code anyway because the iteration feature has been merged.  Will do that when I come back.